Feature/setup wizard

Alapan opened the pull request:
Closes #922 

Important changes:
 
- Created a server method to check the value of a flag called `initialSetupComplete`. If this returns `false`, an initial setup wizard is to be shown.
- The wizard allows an admin to configure settings for the instance.

brylie commented on lib/_config/router.js:
```diff
@@ -68,6 +68,13 @@ onAfterAction = function() {
 };
 
+Router.onBeforeAction(function() {
+  var configRequired = Meteor.call('checkRequiredSettings');
```
Since this code is relying on a server method, it is asynchronous. So, we need to handle the return value from `checkRequiredSettings` in a callback:

```js
// Instead of
// var configRequired = Meteor.call("checkRequiredSettings");
// It should read

// Get a named reference to current router
// for use inside the server method callback
const currentRouter = this;

Meteor.call("checkRequiredSettings", function (error, result) {
  // if error....

  // do something with the result:
  if (result === true) {
    Router.go("settingsWizard);
  } else {
    // call this.next() using the currentRouter variable from above
    currentRouter.next()
  }
});
```

brylie commented on client/layouts/master_layout/master_layout.js:
```diff
@@ -17,5 +17,9 @@ Template.masterLayout.helpers({
     }
+  },
+  initialSetupComplete: function() {
+    var initialSetupComplete = Meteor.call('additionalSetupRequired');
```
This probably will not work, since method calls are asynchronous. Also, consider a different approach here, rather than modifying the master layout.

In effect, redirecting the user to a setup wizard supersedes this alert, and the alert can be displayed in the setup wizard template instead of the master layout.

brylie commented on client/views/settings_wizard/settingsAlert.html:
```diff
@@ -0,0 +1,9 @@
+  <div class="alert alert-warning">
+    <a href="#" class="close" data-dismiss="alert" aria-label="Close">&times;</a>
+    <p>
+      <strong>Additional configuration required!</strong><br>
```
Avoid shouting at the user.

brylie commented on client/views/settings_wizard/settingsAlert.html:
```diff
@@ -0,0 +1,9 @@
+    <a href="#" class="close" data-dismiss="alert" aria-label="Close">&times;</a>
+    <p>
+      <strong>Additional configuration required!</strong><br>
+        Hey there, this instance of <strong>APINF</strong> needs additional configuration. Click <a href="{{pathFor route='settingsWizard'}}">here</a> to continue. You must be an admin to do this!!
```
Again, avoid shouting at the user.

brylie commented on client/views/settings_wizard/wizard/settingsWizard.js:
```diff
@@ -0,0 +1,106 @@
+  'click #save-settings': function() {
+    // when configuration is done, call server method to set initialSetupComplete to true, so that the settings alert is no longer shown
+    Meteor.call("initialSetupCompleteTrue");
+    Router.go("settingsComplete");
```
Call this from within the `initialSetupCompleteTrue` callback, so the method has time to run first.

brylie commented on lib/_config/router.js:
```diff
@@ -38,6 +38,16 @@ var requireAdminRole = function () {
   }
 };
 
+var additionalSetupRequired = function () {
```
Use `const` here.

brylie commented:
Good work so far. Just a couple of minor details to catch at this point.

brylie commented on client/views/settings_wizard/wizard/settingsWizard.js:
```diff
@@ -0,0 +1,106 @@
+  }
+});
+
+AutoForm.addHooks(['settings'], {
```
This is your AutoForm callback context. You should call the server method from within the `onSuccess` callback.

brylie commented on client/views/settings_wizard/wizard/settingsWizard.js:
```diff
@@ -0,0 +1,106 @@
+  },
+  'click #save-settings': function() {
+    // when configuration is done, call server method to set initialSetupComplete to true, so that the settings alert is no longer shown
+    Meteor.call("initialSetupCompleteTrue");
```
Move this method call to the `AutoForm.onSuccess` callback.

Alapan commented:
@brylie I made the changes you suggested. However, I made a slight change in the logic - I have introduced a new field `initialSetupComplete` in the `Settings` collection. Otherwise, a setting created through the program is not persisting across multiple templates. So if `initialSetupComplete` is set to `true` when the relevant method is called through Template 1, the value is becoming `undefined` when it is checked through Template 2. 

I have also removed the alert logic from the master layout file. Currently the alert is not being shown anywhere.

brylie commented:
Ok, make sure that the problem is not related to meteor restarting when editing the project sourcecode.

If you take the collection approach, be sure to create a publication that only publishes the `initialSetupComplete` field from the settings collection document. Otherwise, we may inadvertently publish sensitive information from our settings collection.

Alapan commented:
I made the changes.

brylie commented:
Great! :-) Is this ready for review/merge?

Alapan commented:
@brylie Yes, this is ready for review.

shaliko commented:
@Alapan project stop works for me if I do ```meteor reset``` and try open any page. Looks like infinite loop.

Alapan commented:
@shaliko I made a small change. Could you please test and let me know if you are still getting the same error?

brylie commented:
When starting the project without any settings, e.g. by running `meteor` without the `--settings` flag, I get the following error in the shell:

```
Error: After filtering out keys not in the schema, your object is now empty
```

brylie commented:
When starting Apinf *without* any settings, i.e. without the `--settings ...` flag, I get the following error:

```
Error: After filtering out keys not in the schema, your object is now empty
```

One goal of this task is that Apinf should start with no initial settings, so the Wizard can display.

brylie commented:
It seems like `/server/startup.js` may need significant re-work/review here. Specifically,

- Lines 11 and 24 attempt to insert the `Meteor.settings` object into the settings collection without verifying the structure of that object
  - initially, `Meteor.settings` has the following structure `{ public: {} }`, which doesn't contain any relevant settings
- There are two large sections of that file with duplicate functionality (lines 4-12 and lines 20-24)
  - refactor this code, so that it only runs once
- We should not call `Meteor.call("createApiUmbrellaWeb")` without first verifying that API Umbrella settings have been stored in the `Settings` *collection*
  - I.e. all of code may be refactored to first use the `Settings` collection as opposed to the `Meteor.settings` object
  - The `Meteor.settings` object will be a temporary placeholder for settings, so they can be persisted in the Settings collection
- Any API Umbrella related methods should only run if there are valid API Umbrella settings in the `Settings` collection
  - The *'Sync API Umbrella Users and API Backends'* `SyncedCron` job should only start if API Umbrella settings are present
    - the job, including any sub-methods,  should check the `Settings` collection, rather than the `Meteor.settings` object
  - `Meteor.call("syncApiUmbrellaUsers");` and `Meteor.call("syncApiBackends");` should also only run if API Umbrella settings are stored in the Settings collection.

brylie commented:
We cannot figure out how to get the settings wizard to display.

Alapan commented:
Currently, after running `meteor reset` and then `meteor` (without the settings file), when a user registers successfully, he is redirected to the wizard. However, currently I am facing a problem setting the Host name to `http://localhost:3000` in the wizard, as the application returns an `Unknown validation error`. Setting the host name to something like `http://example.com` works, though. 

Also, even after configuring the settings successfully, the following errors are being seen in the console: 
1. `[TypeError: Cannot read property 'clientId' of undefined]`
2. `[TypeError: Cannot read property 'username' of undefined]`

brylie commented:
Ah, yes. We had the same validation issue, possibly because of the port number.

As a second thought, it is probably unnecessary to provide the HOST value for the Apinf platform, since Meteor is already aware of this value, so lets deprecate that from the setup wizard/settings.

brylie commented:
@frenchbread, what is your recollection of why we add the `host` settings in `config.json`? I think it had something to do with the dashboard.

brylie commented:
@Alapan regarding the 'clientId' and 'username' errors, we need to check the existance of these fields before trying to use them. Try commenting out lines 11 and 24 of `server/startup.js`, as mentioned above.

bajiat commented:
@Alapan and @shaliko Could you do some pair programming on this one to try solve the issues Alapan is facing? Sometimes even just explaining the problem to someone else might already help.  Shaliko is not available 9 May, but should be back at work 10 May.
@brylie said he was just quessing to a certain extent what the problem was.

Alapan commented:
I tried commenting the lines as @brylie suggested, but it didn't help. Please note, there was no problem in running the application - just that those 2 TypeErrors are present after filling in the settings details.

brylie commented:
Ah, I understand now. Are they happening after you submit the form? It may still be a validation problem, where the schema is not passed in the fields it needs for validation.

brylie commented:
@shaliko are you available today to help troubleshoot this task?

shaliko commented:
@brylie Today joining to this task.

brylie commented:
Thanks @shaliko. @Alapan do you have time to work with @shaliko today?

Alapan commented:
I can work on this today late evening. I don't know if @shaliko will be available then. I will try to fix the error and update in a comment the status of my work.

shaliko commented:
@Alapan I will finish work in ~1 hour. Will work with you next time ;)

Alapan commented:
Along with the type errors, after entering the settings successfully, going to the dashboard page, waiting for the charts to appear, then logging out and logging back in, another error is seen on the console:
`Exception while invoking method 'getChartData' TypeError: Cannot read property 'host' of undefined`.

This is when setting the host name to something like `http://www.example.com`. Clearly, a proper host name is necessary. But entering `http://localhost:3000` is giving an unknown validation error

brylie commented:
Ah, yep. That is a known bug. We need to decouple the dashboard from elasticsearch, or make the elasticsearch settings required.

brylie commented:
For reference: https://github.com/apinf/api-umbrella-dashboard/issues/927

brylie commented:
Ping @ccsr @bajiat @ashakunt

ashakunt commented:
@Alapan issue #927 has been resolved by pull request #944 today. Can you check if this issue is still reproducible? 

Quote: https://github.com/apinf/api-umbrella-dashboard/pull/925#issuecomment-218619497

brylie commented:
The exception in `getChartData` may still be there. Please ignore, and finalize the wizard PR.

Alapan commented:
I didn't see the exception now, and the charts were showing after registering, logging out, and logging back in. Please review.

brylie commented:
Great work!

On the first page of the setup wizard, the 'choose file' buttons do not work in Firefox 46. It may be related to the following upstream bug:

https://github.com/yogiben/meteor-autoform-file/issues/104

brylie commented:
One small improvement would be to redirect the user to the dashboard after the wizard, rather than having them click a link. Our router package has a `Router.go()` method that would work here.

brylie commented on client/views/settings_wizard/settingsAlert.html:
```diff
@@ -0,0 +1,9 @@
+    <a href="#" class="close" data-dismiss="alert" aria-label="Close">&times;</a>
+    <p>
+      <strong>Additional configuration required - </strong><br>
+        Hey there, this instance of <strong>APINF</strong> needs additional configuration. Click <a href="{{pathFor route='settingsWizard'}}">here</a> to continue. You must be an admin to do this.
```
Lets remove this text for a couple of reasons:

- it is no longer necessary, since the wizard automatically shows up without the user needing to click
- text like this, with mixed in markup and 'click here' links is not ideal to translate

brylie commented on client/views/settings_wizard/wizard/settingsComplete.html:
```diff
@@ -0,0 +1,5 @@
+<template name="settingsComplete">
+  <div class="alert alert-success">
+    Settings configured successfully! Click <a href="{{pathFor route='dashboard'}}">here </a> to return to your dashboard.
```
Rather than asking the user to 'click here', lets simply redirect to the dashboard using `Router.go()`

brylie commented on client/views/settings_wizard/wizard/settingsWizard.html:
```diff
@@ -0,0 +1,153 @@
+            <div class="col-xs-12">
+              <div class="panel panel-default">
+                <div class="panel-heading">
+                  <h3 class="panel-title">Step 1 - Project Branding</h3>
```
Add i18n token here

brylie commented on client/views/settings_wizard/wizard/settingsWizard.html:
```diff
@@ -0,0 +1,153 @@
+                  <div id="form-buttons">
+                    <div class="row">
+                      <div class="col-xs-12 col-sm-6 form-group">
+                        <button type="button" class="btn btn-default btn-block" id="prev-first-slide">Previous</button>
```
Add i18n token here

brylie commented on client/views/settings_wizard/wizard/settingsWizard.html:
```diff
@@ -0,0 +1,153 @@
+                        <button type="button" class="btn btn-default btn-block" id="prev-first-slide">Previous</button>
+                      </div>
+                      <div class="col-xs-12 col-sm-6 form-group">
+                        <button type="submit" class="btn btn-success btn-block" id="next">Save and Next</button>
```
Add i18n token here

brylie commented on client/views/settings_wizard/wizard/settingsWizard.html:
```diff
@@ -0,0 +1,153 @@
+                  <div id="form-buttons">
+                    <div class="row">
+                      <div class="col-xs-12 col-sm-6 form-group">
+                        <button type="button" class="btn btn-default btn-block" id="prev-first-slide">Previous</button>
```
Add i18n token here

brylie commented on client/views/settings_wizard/wizard/settingsWizard.html:
```diff
@@ -0,0 +1,153 @@
+                        <button type="button" class="btn btn-default btn-block" id="prev-first-slide">Previous</button>
+                      </div>
+                      <div class="col-xs-12 col-sm-6 form-group">
+                        <button type="submit" class="btn btn-success btn-block" id="next">Save and Next</button>
```
Add i18n token here

brylie commented on client/views/settings_wizard/wizard/settingsWizard.html:
```diff
@@ -0,0 +1,153 @@
+              <div class="well">
+              <div class="panel panel-default">
+                <div class="panel-heading">
+                  <h3 class="panel-title">Step 2 - Settings for API Umbrella</h3>
```
Add i18n token here

brylie commented on client/views/settings_wizard/wizard/settingsWizard.html:
```diff
@@ -0,0 +1,153 @@
+              </div>
+              <div class="panel panel-default">
+                <div class="panel-heading">
+                  <h3 class="panel-title">Api Documentation Editor</h3>
```
Add i18n token here

brylie commented on client/views/settings_wizard/wizard/settingsWizard.html:
```diff
@@ -0,0 +1,153 @@
+              </div>
+              <div class="panel panel-default">
+                <div class="panel-heading">
+                  <h3 class="panel-title">Elasticsearch</h3>
```
Add i18n token here

brylie commented on client/views/settings_wizard/wizard/settingsWizard.html:
```diff
@@ -0,0 +1,153 @@
+              </div>
+              <div class="panel panel-default">
+                <div class="panel-heading">
+                  <h3 class="panel-title">Mail</h3>
```
Add i18n token here

brylie commented on client/views/settings_wizard/wizard/settingsWizard.html:
```diff
@@ -0,0 +1,153 @@
+              <div id="form-buttons">
+                <div class="row">
+                  <div class="col-xs-12 col-sm-6 form-group">
+                    <button type="button" class="btn btn-default btn-block" id="prev-second-slide">Previous</button>
```
Add i18n token here

brylie commented on client/views/settings_wizard/wizard/settingsWizard.html:
```diff
@@ -0,0 +1,153 @@
+                    <button type="button" class="btn btn-default btn-block" id="prev-second-slide">Previous</button>
+                  </div>
+                  <div class="col-xs-12 col-sm-6 form-group">
+                    <button type="submit" class="btn btn-success btn-block" id="save-settings">Finish</button>
```
Add i18n token here

brylie commented on client/views/settings_wizard/wizard/settingsWizard.js:
```diff
@@ -0,0 +1,104 @@
+  onSuccess: function () {
+    // Call method to update Meteor.settings
+    Meteor.call('updateMeteorSettings');
+    FlashMessages.sendSuccess('Settings saved.');
```
Add i18n text here here:

```js
// Get success message
const successMessage = TAPi18n.__("setupWizard_successMessage");
```

brylie commented on client/views/settings_wizard/wizard/settingsWizard.js:
```diff
@@ -0,0 +1,104 @@
+    try {
+      Meteor.call("createApiUmbrellaWeb");
+      Meteor.call("initialSetupCompleteTrue", function() {
+        Router.go("settingsComplete");
```
Instead of routing to the `settingsComplete`, simplify our code/process by routing to the Dashboard and showing a success message.

brylie commented on lib/_config/router.js:
```diff
@@ -68,6 +86,10 @@ onAfterAction = function() {
 
+// not sure how to check additionalSetupRequired for all routes
+// the below implementation causes an infinite redirect loop at sign-in
+Router.onBeforeAction(additionalSetupRequired, {except: []});
```
If this code works as expected, update the comment in the line above. Right now, the comment explains that things are not working as expected.

brylie commented on lib/router/main.js:
```diff
@@ -20,6 +20,18 @@ Router.route("/api/:_id/edit", function () {
+    render: "settingsWizard"
+  });
+
+  this.route("settingsComplete", {
```
Lets deprecate this route and the accompanying template. That will simplify the user-flow and our codebase.

brylie commented on server/methods/settings.js:
```diff
@@ -12,5 +12,17 @@ Meteor.methods({
 
     return apiUmbrellaHost;
+  },
+  'isSettingsObjectValidAndEmpty': function() {
```
Consider how this function name reads in an if statement, in plain english:

> if is settings object valid and empty, then do something

Consider how the function will be used, and update the name so it is grammatical. E.g. `settingsObjectIsValidAndEmpty`

> if settings object is valid and empty, then do something

brylie commented on server/publish/settings.js:
```diff
@@ -14,3 +14,8 @@ Meteor.publish("singleSetting", function (setting) {
+
+Meteor.publish('initialSetupComplete', function() {
+  const settings = Settings.findOne();
+  return settings.initialSetupComplete;
```
Publications need to return a database cursor, as far as I recall. This publication returns a string, so may not work as expected. Remove if unused.

brylie commented on server/startup.js:
```diff
@@ -2,60 +2,59 @@ Meteor.startup(function () {
-      // if not insert settings object
-      Settings.insert(settings);
-    }
+    
```
Cleanup unused code.

brylie commented on server/startup.js:
```diff
@@ -2,60 +2,59 @@ Meteor.startup(function () {
+    }*/
+
+    // Check if Settings collection has existing API Umbrella settings
+    if (Settings.findOne) {
```
This looks like a syntax error. It should probably read:

```js
const settings = Settings.findOne();

if (settings) {}
```

brylie commented:
This is really close. Read through the comments, most of which are minor changes.

Be sure of the following:

- remove any unused code, comments, console.log, etc
- read your code in plain English, to check that it is grammatical
- make sure comments are meaningful and accurate; try to describe the intended goal rather than the code itself

Nazarah commented:
Now, if a mandatory field is not filled up while User configs the APIs, there are errors in loading information in the analytic view. Make sure the configuration fields are mandatory, so that the user does not leave them empty.

Nazarah commented:
If users are clicking save by leaving some config fields empty, make sure Save operation does not take place. Instead alert text appears under the fields which are left empty. (CSS alert class texts)

Nazarah commented:
@Alapan: can you please look at the last 2 comments of mine and give your opinion.
Would they be fixed in this PR or we should add them as enhancements?

Alapan commented:
@Nazarah I think it is necessary to let the user know which fields are mandatory (e.g. by having an asterisk besides the field name), not just for submitting the form but also for displaying analytics data. Since this issue is about creating a wizard and not really form validation, maybe it is better to have the latter as an enhancement.

bajiat commented:
@alapan It is ok to make the as enhancements. Can you still check Brylie's comments (2 days ago) so that the PR can be merged?

Alapan commented:
@bajiat @brylie The changes were made. Please let me know of further changes.

frenchbread commented:
@Alapan Please resolve the conflicts.

Alapan commented:
@frenchbread Done.

jykae commented:
@frenchbread As discussed I assign myself to check out this.

frenchbread commented:
Thanks @Alapan @jykae

jykae commented:
@Alapan Reviewing

jykae commented:
@Alapan This works and looks good. I am merging in.

Editor and Mail cannot be enabled because some validation bugs in the form. Discussed with @bajiat We have issue #961 for email and raised now in planning. Editor is then another case what we decide to do with that in future.
